### DISCUSSION TITLE
How can I bind a configuration section to a KeyValuePair<string, string>?

### DISCUSSION THREAD

**Discussion by paulomorgado** [dotnet/runtime#97039] [Q&A]
```csharp
var configuration = new ConfigurationBuilder()
    .AddInMemoryCollection(
    [
        new("kvp:Key", "Key1"),
        new("kvp:Value", "Value1"),
    ])
    .Build();

var kvp = configuration.GetSection("kvp").Get<KeyValuePair<string, string>>();
```

`kvp.Key` and `kvp.Value` are `null`.

**Comment by elgonzo** [comment]
I don't know of a way to directly use KeyValuePair in this situation. Since structs can always be initialized without using a parameterized constructor, the binder foregoes using KeyValuePair's parameterized constructor when instantiating the KeyValuePair instance, but is then unsuccessful in setting the Key/Value properties as they lack setters.

But if your scenario isn't too involved, one possible workaround might be using your own SettableKVP type with settable properties and an implicit/explicit conversion operator which converts a SettableKVP instance to a KeyValuePair instance.
```C#
KeyValuePair<string, string> kvp = configuration.GetSection("kvp").Get<SettableKVP<string, string>>();

public struct SettableKVP<TKey, TValue>
{
    public TKey Key { get; set; }
    public TValue Value { get; set; }

    public static implicit operator KeyValuePair<TKey, TValue>(SettableKVP<TKey, TValue> kvp)
        => new(kvp.Key, kvp.Value);
}
```

**Reply by paulomorgado** [reply]
The problem here is that I don't control whatever types libraries use for they configuration. 😢

**Reply by elgonzo** [reply]
Are you saying that you want to bind your configuration against some 3rd-party type(s) which themselves use KeyValuePair\<TKey,TValue\> properties independently from dictionaries or similar?

If so, it seems you're out of luck, because the implementation of the binder is such that it doesn't use parameterized constructors for value types. In particular, look at the implementation of the [ConfigurationBinder.CreateInstance method](https://github.com/dotnet/runtime/blob/5598dac557c35d264c527e22239f1b33e0489376/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L442): https://github.com/dotnet/runtime/blob/5598dac557c35d264c527e22239f1b33e0489376/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L458-L459

**Reply by aleksa-kuzman** [reply]
You can do something like 

` IEnumerable<KeyValuePair<string,string>> keyValuePairList = configuration.GetSection("kvp").AsEnumerable();`

From there you could get single keyvalue pair via LINQ
